### ISSUE TITLE
Regarding the issue of R-MFDN network

### ISSUE THREAD

**Issue by xiaoran33** [xyyandxyy/IDForge#2]
Hello, I just finished reading the paper "Identity Driven Multimedia Forgery Detection via Reference Assistance" and feel that it is an excellent paper. Can you provide the source code of the R-MFDN network mentioned in the paper? I want to reproduce the experimental data

**Comment by xyyandxyy** [comment]
Hello, since this work was supported by a project by CEC GienTech Technology Co., Ltd., the corresponding model is currently being used during their product promotion phase. Therefore I am sorry that there are no plans for a public release at the moment.

However, all of the components in our method are readily available. Specifically, for the visual modality, we use a pretrained ResNet as an image-level encoder; for the text modality, we use BERT; and for the audio modality, we utilize the Audio Spectrogram Transformer (https://huggingface.co/docs/transformers/model_doc/audio-spectrogram-transformer).  You can easily reproduce our method.

**Comment by xiaoran33** [comment]
论文中提到“Given suspected media data for one identity, Identity-aware contrastive learning aims to learn identity-aware features for each modality by contrasting its feature with that from non-matching identities.”意思是：“在给定一个身份的可疑媒体数据的情况下，身份感知对比学习旨在通过将其特征与非匹配身份的特征进行对比来学习每种模态的身份感知特征。”这里的“非匹配身份”是什么意思呢？没理解

**Comment by xyyandxyy** [comment]
> 论文中提到“Given suspected media data for one identity, Identity-aware contrastive learning aims to learn identity-aware features for each modality by contrasting its feature with that from non-matching identities.”意思是：“在给定一个身份的可疑媒体数据的情况下，身份感知对比学习旨在通过将其特征与非匹配身份的特征进行对比来学习每种模态的身份感知特征。”这里的“非匹配身份”是什么意思呢？没理解

您好, 这里非匹配身份指的是对比学习中的负样本对.

例如针对Bob这个目标人物的身份, 视频A(真实)中Bob的特征
* 和视频B(真实)中Bob的特征互为正样本, 因为`真实`+`同属于一个人`
* 和视频C(伪造)中Bob的特征互为负样本, 应为包含伪造就均为negative.
* 与视频D(真实)中Alice的特征为负样本, 因为不属于同一个人

**Comment by xiaoran33** [comment]
![Image](https://github.com/user-attachments/assets/cf4406e0-9498-40f3-bbfa-8dda20ec9c3f)

![Image](https://github.com/user-attachments/assets/929350eb-c896-425d-9f22-9d60e30a1c0b)


在4.3章节的身份感知对比学习提到：“给定训练集中的ith原始视频镜头样本，通过在参考集中选择具有相同身份的样本来构造其正对。同时，通过在参考集中选择具有不同身份的样本或在训练集中选择操纵样本来构造负对。因此，我们有：上图的损失函数公式。其中，fip表示训练集中ith原始样本的p模态特征；jth样本从参考集中采样，与ith样本具有相同的身份。fkp既可以是来自具有不同身份
的参考集的特征，也可以是来自与ith样本具有相同身份的被操纵的训练数据的特征。通过这种方式，它强制要求编码器学习具有身份感知能力的特性。”   

ith是指的原始视频镜头，为real视频，jth为参考集样本，也是real视频，那都是都real的视频啊？

**Comment by xyyandxyy** [comment]
> ![Image](https://github.com/user-attachments/assets/cf4406e0-9498-40f3-bbfa-8dda20ec9c3f)
> 
> ![Image](https://github.com/user-attachments/assets/929350eb-c896-425d-9f22-9d60e30a1c0b)
> 
> 在4.3章节的身份感知对比学习提到：“给定训练集中的ith原始视频镜头样本，通过在参考集中选择具有相同身份的样本来构造其正对。同时，通过在参考集中选择具有不同身份的样本或在训练集中选择操纵样本来构造负对。因此，我们有：上图的损失函数公式。其中，fip表示训练集中ith原始样本的p模态特征；jth样本从参考集中采样，与ith样本具有相同的身份。fkp既可以是来自具有不同身份 的参考集的特征，也可以是来自与ith样本具有相同身份的被操纵的训练数据的特征。通过这种方式，它强制要求编码器学习具有身份感知能力的特性。”
> 
> ith是指的原始视频镜头，为real视频，jth为参考集样本，也是real视频，那都是都real的视频啊？

您说`ith是指的原始视频镜头，为real视频，jth为参考集样本，也是real视频，那都是都real的视频啊？`, 这里$f_i$和$f_j$确实都是来源于真实的视频的特征, 因此两者才作为正样本放在分子项.  具体可以看一下InfoNCE Loss的定义.

**Comment by xiaoran33** [comment]
好的，感谢解答。
```
|-- face_audiomismatch_textmismatch
|-- face_rvc_textmismatch
|-- face_tts
|-- face_tts_textgen
|-- lip_audiomismatch_textmismatch
|-- lip_rvc_textmismatch
|-- lip_tts_textgen
|-- pristine
|-- rvc_textmismatch
|-- tts_textgen
`-- tts_textmismatch ##### Forgery Type
    |-- id01
    |-- id02
    |-- ...
    `-- id53 ##### Person
        |-- id53_00
        |-- id53_04
        |-- id53_05
        `-- id53_08 ##### Different Source Video
            |-- id53_scene_0083_0_simswap.mp4
            |-- id53_scene_0090_0_infoswap.mp4
            |-- id53_scene_0090_0_roop.mp4
            |-- id53_scene_0090_1_simswap.mp4
            `-- id53_scene_0092_0_roop.mp4 ##### Video File
                |-- frames_ndarray.npy ##### Extracted Frame (ndarray format)
                |-- id53_scene_0092_0_roop.mp3  ##### Extracted Audio by FFmpeg
                `-- id53_scene_0092_0_roop.txt ##### Extracted Transcript by Whisper
```
再咨询一个问题：就是github上提到IDForge v2是我们论文中使用的版本，其中每个预处理的视频保留16帧（以相等的间隔分为4组采样，每组4帧）、音频文件和提取的文本。”，怎么理解“每个预处理的视频保留16帧（以相等的间隔分为4组采样，每组4帧”，一帧我理解为一个图像，一个图像就是一个图片吧，那怎么和音频文件和转录文本对齐的？frames_ndarray.npy是16帧的数据吗？id53_scene_0092_0_roop.mp3和 id53_scene_0092_0_roop.txt是连续的吧？一个离散和怎么和两个连续的是对齐的呢？

**Comment by xyyandxyy** [comment]
> 好的，感谢解答。
> 
> ```
> |-- face_audiomismatch_textmismatch
> |-- face_rvc_textmismatch
> |-- face_tts
> |-- face_tts_textgen
> |-- lip_audiomismatch_textmismatch
> |-- lip_rvc_textmismatch
> |-- lip_tts_textgen
> |-- pristine
> |-- rvc_textmismatch
> |-- tts_textgen
> `-- tts_textmismatch ##### Forgery Type
>     |-- id01
>     |-- id02
>     |-- ...
>     `-- id53 ##### Person
>         |-- id53_00
>         |-- id53_04
>         |-- id53_05
>         `-- id53_08 ##### Different Source Video
>             |-- id53_scene_0083_0_simswap.mp4
>             |-- id53_scene_0090_0_infoswap.mp4
>             |-- id53_scene_0090_0_roop.mp4
>             |-- id53_scene_0090_1_simswap.mp4
>             `-- id53_scene_0092_0_roop.mp4 ##### Video File
>                 |-- frames_ndarray.npy ##### Extracted Frame (ndarray format)
>                 |-- id53_scene_0092_0_roop.mp3  ##### Extracted Audio by FFmpeg
>                 `-- id53_scene_0092_0_roop.txt ##### Extracted Transcript by Whisper
> ```
> 
> 再咨询一个问题：就是github上提到IDForge v2是我们论文中使用的版本，其中每个预处理的视频保留16帧（以相等的间隔分为4组采样，每组4帧）、音频文件和提取的文本。”，怎么理解“每个预处理的视频保留16帧（以相等的间隔分为4组采样，每组4帧”，一帧我理解为一个图像，一个图像就是一个图片吧，那怎么和音频文件和转录文本对齐的？frames_ndarray.npy是16帧的数据吗？id53_scene_0092_0_roop.mp3和 id53_scene_0092_0_roop.txt是连续的吧？一个离散和怎么和两个连续的是对齐的呢？

frames_ndarray.npy是16帧的数据, id53_scene_0092_0_roop.mp3和 id53_scene_0092_0_roop.txt是连续的.

视频采取部分帧是常见操作, 因为GPU显存限制, 通常是没办法采大量帧放进显存中. 之前的工作通常是均匀采样, 比如一个1000帧里面均匀步长采样16帧, 而我们这边是对1000帧按均匀的步长采样4组, 每组采样连续的4帧.

关于对齐的问题, 我们是让模型通过训练自己去学习如何对其不同模态数据在特征空间的特征.

**Comment by xiaoran33** [comment]
![Image](https://github.com/user-attachments/assets/dad7c5f8-d483-4187-afbc-802ce3740651)

自己去学习指的是 4.1多模态特征学习这个章节吗？就是使用自我注意力、多头注意力、前馈网络的机制来学习的吗？

**Comment by xyyandxyy** [comment]
> ![Image](https://github.com/user-attachments/assets/dad7c5f8-d483-4187-afbc-802ce3740651)
> 
> 自己去学习指的是 4.1多模态特征学习这个章节吗？就是使用自我注意力、多头注意力、前馈网络的机制来学习的吗？

  你可以这样理解, 你提到的`自我注意力、多头注意力、前馈网络的机制`就是我们模型用来处理特征并且融合特征的方法, 为了能找到这些`自我注意力、多头注意力、前馈网络的机制`的合适参数(使之对齐不同模态数据在特征空间的特征), 我们主要通过的cross-modal contrastive learning loss function (论文中的公式4)来对模型的训练进行约束. 

当然了, 公式4只是我们文中的损失函数之一, 因为我们还考虑了其他训练目标(例如文中section 4.3的Identity-aware Contrastive Learning), 我们整个训练过程用的是完整的损失函数(文中公式9).

**Comment by xiaoran33** [comment]
好的，感谢解答。
上面提到的目录结构中：
 |-- id53_scene_0090_0_roop.mp4
 |-- id53_scene_0090_1_simswap.mp4
“scene”、“0090”、“1”分别代表什么意思呢？

**Comment by xiaoran33** [comment]
好的，感谢解答。
上面提到的目录结构中：
 |-- id53_scene_0090_0_roop.mp4
 |-- id53_scene_0090_1_simswap.mp4
“scene”、“0090”、“1”分别代表什么意思呢？

**Comment by xyyandxyy** [comment]
> 好的，感谢解答。 上面提到的目录结构中： |-- id53_scene_0090_0_roop.mp4 |-- id53_scene_0090_1_simswap.mp4 “scene”、“0090”、“1”分别代表什么意思呢？

这里是我们数据合成过程中的中间标签(例如用了第几张source image, 第几个target video), 在您使用的场景中可能不具备太大参考价值.
